File: beginner0107/DFS/여행경로.java
import java.util.*;

class 여행경로 {
    public List<String> solution(String[][] tickets) {
        Map<String, List<String>> graph = new HashMap<>();
        for (String[] ticket : tickets) {
            String start = ticket[0];
            String end = ticket[1];
            if (graph.containsKey(start)) {
                graph.get(start).add(end);
            } else {
                List<String> destinations = new ArrayList<>();
                destinations.add(end);
                graph.put(start, destinations);
            }
        }

        for (List<String> destinations : graph.values()) {
            Collections.sort(destinations, Collections.reverseOrder());
        }

        List<String> answer = new ArrayList<>();
        Deque<String> stack = new ArrayDeque<>();
        stack.push("ICN");

        while (!stack.isEmpty()) {
            String node = stack.peek();
            if (!graph.containsKey(node) || graph.get(node).isEmpty()) {
                answer.add(stack.pop());
            } else {
                String next = graph.get(node).remove(graph.get(node).size() - 1);
                stack.push(next);
            }
        }

        Collections.reverse(answer);
        return answer;
    }
}

File: beginner0107/DFS/여행경로.py
from collections import deque

def solution(tickets):
    graph = {}
    for start, end in tickets:
        if start in graph:
            graph[start].append(end)
        else:
            graph[start] = [end]
    
    for key in graph:
        graph[key].sort(reverse=True)
    
    stack = ["ICN"]
    answer = []
    while stack:
        node = stack[-1]
        if node not in graph or not graph[node]:
            answer.append(stack.pop())
        else:
            stack.append(graph[node].pop())

    return answer[::-1]
